digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/ProjectLinkSplitter.scala

```diff
@@ -0,0 +1,216 @@
+    def findMatchingSegment(suravageGeom: Seq[Point], templateGeom: Seq[Point]): Option[Seq[Point]] = {
+      if (GeometryUtils.areAdjacent(suravageGeom.head, templateGeom.head, MaxDistanceForConnectedLinks)) {
+        val boundaries = geometryToBoundaries(suravageGeom)
+        val templateSegments = GeometryUtils.geometryToSegments(templateGeom)
+        val exitPoint = templateSegments.flatMap { seg =>
+          findIntersection(seg, boundaries._1, Some(Epsilon), Some(Epsilon))
+            .orElse(findIntersection(seg, boundaries._2, Some(Epsilon), Some(Epsilon)))
+        }.headOption
+        if (exitPoint.nonEmpty) {
+          // Exits the tunnel -> find point
+          exitPoint.map(ep => GeometryUtils.truncateGeometry2D(templateGeom, 0.0,
+            GeometryUtils.calculateLinearReferenceFromPoint(ep, templateGeom)))
+        } else {
+          val length = GeometryUtils.geometryLength(suravage.geometry)
+          Some(GeometryUtils.truncateGeometry2D(templateGeom, 0.0,
+            Math.max(GeometryUtils.geometryLength(template.geometry), length)))
+        }
+      } else if (GeometryUtils.areAdjacent(suravageGeom.last, templateGeom.head, MaxDistanceForConnectedLinks)) {
+        findMatchingSegment(suravageGeom.reverse, templateGeom)
+      } else
+        None
+    }
+    findMatchingSegment(suravage.geometry, template.geometry).orElse(
+      findMatchingSegment(suravage.geometry.reverse, template.geometry.reverse).map(_.reverse))
+  }
+
+  def findIntersection(geometry1: Seq[Point], geometry2: Seq[Point], maxDistance1: Option[Double] = None,
+                       maxDistance2: Option[Double] = None): Option[Point] = {
+    val segments1 = geometry1.zip(geometry1.tail)
+    val segments2 = geometry2.zip(geometry2.tail)
+    val s = segments1.flatMap( s1 => segments2.flatMap{ s2 =>
+      intersectionPoint(s1, s2).filter(p =>
+        maxDistance1.forall(d => GeometryUtils.minimumDistance(p, s1) < d) &&
+          maxDistance2.forall(d => GeometryUtils.minimumDistance(p, s2) < d))})
+    s.headOption
+  }
+
+  def intersectionPoint(segment1: (Point, Point), segment2: (Point, Point)): Option[Point] = {
+
+    def switchXY(p: Point) = {
+      Point(p.y, p.x, p.z)
+    }
+    val ((p1,p2),(p3,p4)) = (segment1, segment2)
+    val v1 = p2-p1
+    val v2 = p4-p3
+    if (Math.abs(v1.x) < 0.001) {
+      if (Math.abs(v2.x) < 0.001) {
+        // Both are vertical or near vertical -> swap x and y and recalculate
+        return intersectionPoint((switchXY(p1),switchXY(p2)), (switchXY(p3), switchXY(p4))).map(switchXY)
+      } else {
+        val dx = (p1 - p3).x
+        val normV2 = v2.normalize2D()
+        return Some(p3 + normV2.scale(dx/normV2.x))
+      }
+    } else if (Math.abs(v2.x) < 0.001) {
+      // second parameter is near vertical, switch places and rerun
+      return intersectionPoint((switchXY(p3), switchXY(p4)), (switchXY(p1),switchXY(p2))).map(switchXY)
+    }
+
+    val a = v1.y / v1.x
```

Comment:
Please do review the variable names of this method (intersectionPoint if I am not mistaken) they are very non intuitive.